src/2d-map.jl

```diff
@@ -0,0 +1,302 @@
+
+
+struct Halo2DTiling <: FetchingScheme
+    depth::Int
+    halo::Float64
+    pixel_scale::Float64
+end
+
+Halo2DTiling(; depth=8, halo=0.2, pixel_scale=2.0) = Halo2DTiling(depth, halo, pixel_scale)
+
+function tile_key(provider::AbstractProvider, tile::Tile)
+    return TileProviders.geturl(provider, tile.x, tile.y, tile.z)
+end
+
+"""
+    Map
+
+    Map(extent, [extent_crs=wgs84]; kw...)
+
+Tylers main object, it plots tiles onto a Makie.jl `Axis`,
+downloading and plotting more tiles as you zoom and pan.
+
+# Arguments
+
+-`extent`: the initial extent of the map, as a `GeometryBasics.Rect`
+    or an `Extents.Extent` in the projection of `extent_crs`.
+-`extent_crs`: Any `GeoFormatTypes` compatible crs, the default is wsg84.
+
+# Keywords
+
+-`resolution`: The figure resolution.
+-`figure`: an existing `Makie.Figure` object.
+-`crs`: The providers coordinate reference system.
+-`provider`: a TileProviders.jl `Provider`.
+-`max_parallel_downloads`: limits the attempted simultaneous downloads, with a default of `16`.
+-`cache_size_gb`: limits the cache for storing tiles, with a default of `5`.
+-`depth`: the number of layers to load when zooming. Lower numbers will be slightly faster
+    but have more artefacts. The default is `8`.
+-`halo`: The fraction of the width of tiles to add as a halo so that panning is smooth - the
+    tiles will already be loaded. The default is `0.2`, which means `0.1` on each side.
+-`scale`: a tile scaling factor. Low number decrease the downloads but reduce the resolution.
+    The default is `1.0`.
+"""
+struct Map{Ax<:Makie.AbstractAxis} <: AbstractMap
+    provider::AbstractProvider
+    figure::Figure
+    axis::Ax
+    plot_config::AbstractPlotConfig
+    # The tile downloader + cacher
+    tiles::TileCache
+    # The tiles for the current zoom level - we may plot many more than this
+    current_tiles::ThreadSafeDict{Tile, Bool}
+    # The plots we have created but are not currently visible and can be reused
+    unused_plots::Vector{Makie.Plot}
+    # All tile plots we're currently plotting
+    plots::ThreadSafeDict{String,Tuple{Makie.Plot,Tile, Rect}}
+    # All tiles we currently wish to be plotting, but may not yet be downloaded + displayed
+    should_be_plotted::ThreadSafeDict{String, Tile}
+    display_task::Base.RefValue{Task}
+
+    crs::CoordinateReferenceSystemFormat
+    zoom::Observable{Int}
+    fetching_scheme::FetchingScheme
+    max_zoom::Int
+    max_plots::Int
+end
+
+setup_axis!(::Makie.AbstractAxis, ext_target) = nothing
+
+function setup_axis!(axis::Axis, ext_target)
```

Comment:
Should we also pass the extent CRS here?  GeoAxis might need that, since a user might pass some e.g. `PolarStereographic` crs for the extent!

Reply:
You mean the target crs?

Reply:
Yeah, the CRS of `extent_target` basically.